### DISCUSSION TITLE
Cypher Collect

### DISCUSSION THREAD

**Discussion by JoelLeupp** [kuzudb/kuzu#1166] [General]
I want aggregate properties of a node into a list and for that case I wanted to use the cypher collect function like in the example here: [neo4s](https://neo4j.com/developer/cypher/controlling-query-processing/) and write a query like:

MATCH (p:Person)-[:IS_FRIENDS_WITH]->(friend:Person)
WHERE size((friend)-[:IS_FRIENDS_WITH]-(:Person)) > 1
RETURN p.name, collect(friend.name) AS friends, size((friend)-[:IS_FRIENDS_WITH]-(:Person)) AS numberOfFoFs;

But when I try to execute that query I get this error: RuntimeError: Catalog exception: COLLECT function does not exist.
I also didn't find the function in the kuzu docs so I guess it is not implemented. Is there a way to work around that or is it planed that the collect function will be available soon?

Thanks for the answer

**Comment by andyfengHKU** [comment]
Hi,

Thanks for trying us out. Currently neither size() nor collect() are implemented. As far as I know, there is no alternative cypher statement to achieve the same effect of collect(). 

For collect() function, we didn't prioritize it but we can implement probably in 1-2 weeks.

For size() function, it's tricky to support since we don't have notion of PATH in the system yet. But if my understanding is correct, you seems to be asking for the degree of friend. We can also try to support degree function in a few weeks.

**Reply by JoelLeupp** [reply]
Hi, thanks for the fast answer for me collect is the important function. My concrete use-case is that I have a collaboration graph where the edges are collaborations between authors and the edge has as a property the id of the paper in which they collaborated.  My goal is to get weighted edges with a list of to property id of the aggregated edges For this I would aggregate the edges (collaboration) with the function count() to get the weight and aggregate with collect() to get a list of all paper ids of which to authors collaborated with each other.

A simple example would be to go from this output:

| Author_X | Author_Y | Paper_ID |
|----------|----------|----------|
| A        | B        | 111      |
| A        | C        | 222      |
| A        | B        | 333      |
| A        | B        | 444      |

To this goal output:

| Author_X | Author_Y | Weight | IDs           |
|----------|----------|--------|---------------|
| A        | B        | 3      | [111,333,444] |
| A        | C        | 1      | [222]         |

I want to get weighted edges and keep the information of the papers that they collaborated in.
It would be very cool if such queries will get supported.

**Reply by andyfengHKU** [reply]
I think you are right that collect() is an important function in your use case. Your case will require a cypher statement along the line of 
```
MATCH (x:Author)-[e:Collaborates]->(y:Author) RETURN x.ID, y.ID, COUNT(*) AS weight, COLLECT(e.paper_ID) AS IDs
```

On the other hand, size() function seems to be irrelevant. 

In such case, we will try to get COLLECT() in as fast as we can.

**Reply by JoelLeupp** [reply]
Exactly, I do not need size() just collect and the query would look like you suggested. 

Thanks a lot that would be perfect! I'm looking forward to it.

**Comment by acquamarin** [comment]
Hi,
We have implemented `COLLECT()` aggregate function in #1292. You may need to rebuild `Kùzu` from master so you can get all new features. If you find any bug, feel free to report to us.
